### PR TITLE
Implement the required NvAPI entrypoints for D3D11 DLSS support.

### PR DESCRIPTION
**Pull request by adamnv**
Note: To initialize [DLSS](https://developer.nvidia.com/dlss) successfully at runtime this also needs some
corresponding DXVK [changes](https://github.com/doitsujin/dxvk/pull/2260).

### CODE REVIEW

File: src/d3d11/nvapi_d3d11_device.h
Comment (jp7677) [inline src/d3d11/nvapi_d3d11_device.h:19]: See above regarding formatting.

File: src/d3d11/nvapi_d3d11_device.h
Comment (jp7677) [inline src/d3d11/nvapi_d3d11_device.h:22]: See above regarding formatting.

File: src/d3d11/nvapi_d3d11_device.h
Comment (jp7677) [inline src/d3d11/nvapi_d3d11_device.h:34]: See above regarding formatting and method signature.

File: src/nvapi_d3d.cpp
Comment (jp7677) [inline src/nvapi_d3d.cpp:9]: Oh nice, thanks, I didn't know about `__FUNCTION__`.  Could you please keep these new methods consistent and use the method name typed out? Additionally please open an issue (or a PR:)) for replacing those typed out function names everywhere with `__FUNCTION__`.

File: src/nvapi_d3d.cpp
Comment (jp7677) [inline src/nvapi_d3d.cpp:13]: Please use cpp style casts.

File: src/nvapi_d3d.cpp
Comment (jp7677) [inline src/nvapi_d3d.cpp:15]: Since `n` is used only once, we can omit the declaration and use the value directly like `return Ok("NvAPI_D3D_GetObjectHandleForResource", alreadyLogged);`. (See comment above about `__FUNCTION__`)

File: src/nvapi_d3d.cpp
Comment (jp7677) [inline src/nvapi_d3d.cpp:27]: Since `n` is used only once, we can omit the declaration and use the value directly like `return Ok("NvAPI_D3D_BeginResourceRendering", alreadyLogged);`. (See comment above about `__FUNCTION__`)

File: src/nvapi_d3d.cpp
Comment (jp7677) [inline src/nvapi_d3d.cpp:34]: Since `n` is used only once, we can omit the declaration and use the value directly like `return Ok("NvAPI_D3D_EndResourceRendering", alreadyLogged);`. (See comment above about `__FUNCTION__`)

File: src/nvapi_d3d.cpp
Comment (jp7677) [inline src/nvapi_d3d.cpp:64]: Please format the `switch` like here: https://github.com/jp7677/dxvk-nvapi/blob/master/src/nvapi_gpu.cpp#L195

File: src/nvapi_d3d.cpp
Comment (jp7677) [inline src/nvapi_d3d.cpp:63]: See comment above about `__FUNCTION__`.

File: src/nvapi_d3d11.cpp
Comment (jp7677) [inline src/nvapi_d3d11.cpp:93]: See comment above about `__FUNCTION__`.

File: src/nvapi_d3d11.cpp
Comment (jp7677) [inline src/nvapi_d3d11.cpp:105]: See comment above about `__FUNCTION__`.

File: src/nvapi_d3d11.cpp
Comment (jp7677) [inline src/nvapi_d3d11.cpp:117]: See comment above about `__FUNCTION__`.

File: src/nvapi_d3d11.cpp
Comment (jp7677) [inline src/nvapi_d3d11.cpp:129]: See comment above about `__FUNCTION__`.

File: src/nvapi_d3d11.cpp
Comment (jp7677) [inline src/nvapi_d3d11.cpp:141]: See comment above about `__FUNCTION__`.

File: src/nvapi_d3d11.cpp
Comment (jp7677) [inline src/nvapi_d3d11.cpp:153]: See comment above about `__FUNCTION__`.

File: src/nvapi_d3d11.cpp
Comment (jp7677) [inline src/nvapi_d3d11.cpp:165]: See comment above about `__FUNCTION__`.

File: src/nvapi_d3d11.cpp
Comment (jp7677) [inline src/nvapi_d3d11.cpp:177]: See comment above about `__FUNCTION__`.

File: src/nvapi_d3d11.cpp
Comment (jp7677) [inline src/nvapi_d3d11.cpp:189]: See comment above about `__FUNCTION__`.

File: src/nvapi_d3d11.cpp
Comment (jp7677) [inline src/nvapi_d3d11.cpp:202]: See comment above about `__FUNCTION__`.

File: src/nvapi_d3d11.cpp
Comment (jp7677) [inline src/nvapi_d3d11.cpp:220]: See comment above about `__FUNCTION__`.

File: src/nvapi_d3d11.cpp
Comment (jp7677) [inline src/nvapi_d3d11.cpp:232]: See comment above about `__FUNCTION__`.

File: src/nvapi_d3d.cpp
Comment (jp7677) [inline src/nvapi_d3d.cpp:75]: See above about the `switch` formatting.

File: src/nvapi_d3d.cpp
Comment (jp7677) [inline src/nvapi_d3d.cpp:79]: Could you please use cpp style casts?

File: src/nvapi_d3d.cpp
Comment (jp7677) [inline src/nvapi_d3d.cpp:87]: Could you please use cpp style casts?

File: src/nvapi_interface.cpp
Comment (jp7677) [inline src/nvapi_interface.cpp:76]: Could please put the D3D11 methods below other D3D11 methods above?

File: src/nvapi_interface.cpp
Comment (jp7677) [inline src/nvapi_interface.cpp:84]: Could please put the D3D methods below other D3D methods above?

File: src/nvapi_interface.cpp
Comment (jp7677) [inline src/nvapi_interface.cpp:89]: Could please put this D3D1x methods below other D3D11 methods above?

**Review by jp7677 — CHANGES_REQUESTED** [review]
Hi @adamnv, thanks a lot for the PR, very cool! The new code looks pretty good, mostly minor suggestions and formatting.

Do you know how often those D3D calls are made into DXVK? I know that UE4 may call `SetDepthBoundsTests` like 500 times a second, hence I keep a reference to the DXVK context stored here to avoid getting the context over and over again (https://github.com/jp7677/dxvk-nvapi/blob/master/src/d3d11/nvapi_d3d11_device.cpp#L5). Does something like this makes sense for those new methods? Note though that doing so violates the COM reference  counting, but based on what I understand from the nature of a context that shouldn't be an issue.

Regarding tests, well, could you please also extend the D3D11 tests with the new methods, see https://github.com/jp7677/dxvk-nvapi/blob/master/tests/nvapi_d3d11.cpp#L10? The tests are my safety net against any regression when eventually doing some internal restructuring.

**Comment by jp7677** [comment]
@adamnv I have approved CI, could you please also look at the now failing tests?

**Comment by jp7677** [comment]
By the way, the committer on your commit is a different user than the author, this is on purpose?

**Comment by adamnv** [comment]
I'm in the process of moving a bunch of the corresponding DXVK code around which will temporarily break the changes in this PR - hopefully I'll fix this PR shortly after!

**Comment by adamnv** [comment]
-> Converted to a draft PR while this requires some outstanding changes to match the new interface updates in https://github.com/doitsujin/dxvk/pull/2260

**Comment by adamnv** [comment]
This is compatible with DXVK PR https://github.com/doitsujin/dxvk/pull/2260 again now and (in theory) it's safe for the corresponding DXVK and dxvk-nvapi to be deployed out of lockstep.

File: tests/nvapi_d3d11_mocks.cpp
Comment (adamnv) [inline tests/nvapi_d3d11_mocks.cpp:178]: There are expected to be broken right now...

Comment (jp7677) [reply tests/nvapi_d3d11_mocks.cpp:178]: Thanks for the heads up, perfectly cool. Let me know when you want me to look at your work again or when I can somehow assist.
(Unfortunately I have to approve CI runs manually every time when you push :( )

Comment (adamnv) [reply tests/nvapi_d3d11_mocks.cpp]: D'oh!  Well, thanks for doing it!

**Comment by jp7677** [comment]
Hi @adamnv I just pushed some changes to nvapi-d3d11 on master to make those part more in line with d3d12 and to remove some of the code duplication. Unfortunately this will conflict now with your work. I hope you won't a hard time resolving this and hopefully this makes things also a bit easier for you.

**Comment by adamnv** [comment]
Thanks for the heads-up.
About the tests - I haven't much looked at why the checks are failing yet, let alone added new tests.  My glance at the check results were a bit baffling, about some code expecting to be hit X times and were hit 0 times?  Any great ideas about that?  I'll get to looking though.

**Comment by jp7677** [comment]
With regards to the tests, may be there is an issue with the way you resolve things from the cache map? E.g. the last error: `Expected context.MultiDrawIndirect(drawCount, &buffer, offsetForArgs,
  strideForArgs) to be called once, actually never called` from https://github.com/jp7677/dxvk-nvapi/blob/master/tests/nvapi_d3d11.cpp#L166 says: when executing `NvAPI_D3D11_MultiDrawIndexedInstancedIndirect` I expected `MultiDrawIndirect` on the provided `context` to be called, but it hasn't been called.

**Comment by adamnv** [comment]
I *think* I just have new tests remaining to finish, maybe some minor reformatting, then GTG?

**Comment by adamnv** [comment]
I've tested that the combination of dxvk-with-dlss and dxvknvapi-without-dlss doesn't fail in unexpected ways, I'll test that dxvk-without-dlss and dxvknvapi-with-dlss also works.

File: src/d3d11/nvapi_d3d11_device.cpp
Comment (jp7677) [inline src/d3d11/nvapi_d3d11_device.cpp:1]: Please remove that empty line.

File: src/d3d11/nvapi_d3d11_device.cpp
Comment (jp7677) [inline src/d3d11/nvapi_d3d11_device.cpp:56]: No need for a separate variable, lets return directly.

File: src/d3d11/nvapi_d3d11_device.cpp
Comment (jp7677) [inline src/d3d11/nvapi_d3d11_device.cpp:72]: Please add an empty line after the `then`  of the if statement.

File: src/d3d11/nvapi_d3d11_device.cpp
Comment (jp7677) [inline src/d3d11/nvapi_d3d11_device.cpp:147]: Please add an empty line after the `then`  of the if statement.

File: src/d3d11/nvapi_d3d11_device.cpp
Comment (jp7677) [inline src/d3d11/nvapi_d3d11_device.cpp:150]: Please add an empty line after the `then`  of the if statement.

File: src/d3d11/nvapi_d3d11_device.cpp
Comment (jp7677) [inline src/d3d11/nvapi_d3d11_device.cpp:158]: Please add an empty line after the `then`  of the if statement.

File: src/d3d11/nvapi_d3d11_device.cpp
Comment (jp7677) [inline src/d3d11/nvapi_d3d11_device.cpp:161]: Please add an empty line after the `then`  of the if statement.

**Review by jp7677 — CHANGES_REQUESTED** [review]

**Review by jp7677 — APPROVED** [review]

**Pull request merged**
